storage/spanner.go

```diff
@@ -0,0 +1,984 @@
+	table = changeTableNameForSP(table)
+	for i := 0; i < len(m); i++ {
+		for k, v := range m[i] {
+			t, ok := ddl[k]
+			if t == "BYTES(MAX)" && ok {
+				ba, err := json.Marshal(v)
+				if err != nil {
+					return errors.New("ValidationException", err)
+				}
+				m[i][k] = ba
+			}
+		}
+		mutations[i] = spanner.InsertOrUpdateMap(table, m[i])
+	}
+	_, err := s.getSpannerClient(table).Apply(ctx, mutations)
+	if err != nil {
+		return errors.New("ResourceNotFoundException", err.Error())
+	}
+	return nil
+}
+
+// SpannerDelete - this will delete the data
+func (s Storage) SpannerDelete(ctx context.Context, table string, m map[string]interface{}, eval *models.Eval, expr *models.UpdateExpressionCondition) error {
+	_, err := s.getSpannerClient(table).ReadWriteTransaction(ctx, func(ctx context.Context, t *spanner.ReadWriteTransaction) error {
+		tmpMap := map[string]interface{}{}
+		for k, v := range m {
+			tmpMap[k] = v
+		}
+		if len(eval.Attributes) > 0 || expr != nil {
+			status, err := evaluateConditionalExpression(ctx, t, table, tmpMap, eval, expr)
+			if err != nil {
+				return err
+			}
+			if !status {
+				return errors.New("ConditionalCheckFailedException", tmpMap, expr)
+			}
+		}
+		tableConf, err := config.GetTableConf(table)
+		if err != nil {
+			return err
+		}
+		table = changeTableNameForSP(table)
+
+		pKey := tableConf.PartitionKey
+		pValue, ok := tmpMap[pKey]
+		if !ok {
+			return errors.New("ResourceNotFoundException", pKey)
+		}
+		var key spanner.Key
+		sKey := tableConf.SortKey
+		if sKey != "" {
+			sValue, ok := tmpMap[sKey]
+			if !ok {
+				return errors.New("ResourceNotFoundException", pKey)
+			}
+			key = spanner.Key{pValue, sValue}
+
+		} else {
+			key = spanner.Key{pValue}
+		}
+
+		mutation := spanner.Delete(table, key)
+		err = t.BufferWrite([]*spanner.Mutation{mutation})
+		if e := errors.AssignError(err); e != nil {
+			return e
+		}
+		return nil
+	})
+	return err
+}
+
+// SpannerBatchDelete - this delete the data in batch
+func (s Storage) SpannerBatchDelete(ctx context.Context, table string, keys []map[string]interface{}) error {
+	tableConf, err := config.GetTableConf(table)
+	if err != nil {
+		return err
+	}
+	table = changeTableNameForSP(table)
+
+	pKey := tableConf.PartitionKey
+	ms := make([]*spanner.Mutation, len(keys))
+	sKey := tableConf.SortKey
+	for i := 0; i < len(keys); i++ {
+		m := keys[i]
+		pValue, ok := m[pKey]
+		if !ok {
+			return errors.New("ResourceNotFoundException", pKey)
+		}
+		var key spanner.Key
+		if sKey != "" {
+			sValue, ok := m[sKey]
+			if !ok {
+				return errors.New("ResourceNotFoundException", sKey)
+			}
+			key = spanner.Key{pValue, sValue}
+
+		} else {
+			key = spanner.Key{pValue}
+		}
+		ms[i] = spanner.Delete(table, key)
+	}
+	_, err = s.getSpannerClient(table).Apply(ctx, ms)
+	if err != nil {
+		return errors.New("ResourceNotFoundException", err)
+	}
+	return nil
+}
+
+// SpannerAdd - Spanner Add functionality like update attribute
```

Comment:
I don't quite understand what is meant by `Add functionality like update attribute`. Looking at the code, it seems that it is intended to add a record (item), and not an attribute.

Reply:
Looking through it again, I think I understand what the intention is: 'Add' an attribute to an existing row (which in Spanner terms means setting the value of a column of an existing row). If that is correct, I would suggest renaming this method to `SpannerUpdate` and also use an `Update` mutation instead of an `InsertOrUpdate` mutation.

Reply:
@rosspatil please look into this